### PR TITLE
Add calculation of alarm times

### PR DESCRIPTION
**Pull request by niccokunzmann**
This contributes to #716 by adding an `icalendar.Alarms` module.
This module allows calculating the alarm times.
It has been tested with

- RFC 9074
- Google
- Outlook
- Nextcloud
- Thunderbird
- Etar

TODO:

- [x] simple interface in Todo and Event
- [x] fix `# TODO:` comments
- [x] add changelog entry
- [x] implement tests that are todo
- [x] access relative alarm times from start and end as property of Alarms

----
📚 Documentation preview 📚: https://icalendar--737.org.readthedocs.build/

### CODE REVIEW

**Comment by niccokunzmann** [comment]
@stevepiercy, could you review the docs?
@jacadzaca, do you have time to have a look at this?

**Comment by stevepiercy** [comment]
@niccokunzmann FYI, I'm backlogged and slammed with preparation for [PloneConf 2024](https://2024.ploneconf.org/en). I estimate it will be at least a week until I can review icalendar PRs.

File: CHANGES.rst [hunk 1]
```diff
@@ -19,6 +21,13 @@ New features:
 
 - Add ``VALARM`` properties for :rfc:`9074`. See `Issue 657 <https://github.com/collective/icalendar/issues/657>`_
 - Test compatibility with Python 3.13
+- Add ``icalendar.alarms`` module to calculate alarm times. See `Issue 716 <https://github.com/collective/icalendar/issues/716>`_.
+- Add ``Event.alarms`` and ``Todo.alarms`` to access alarm calculation.
+- Add ``Component.DTSTAMP`` and ``Component.LAST_MODIFIED`` properties for datetime in UTC.
+- Add ``Component.is_thunderbird()`` to check if the component uses custom properties by Thunderbird.
+- Add ``X_MOZ_SNOOZE_TIME`` and ``X_MOZ_LASTACK`` properties to ``Event`` and ``Todo``.
+- Add ``Alarm.ACKNOWLEDGED``, ``Alarm.TRIGGER``, ``Alarm.REPEAT`` and ``Alarm.DURATION`` properties
```
Comment (stevepiercy) [inline CHANGES.rst]:
```suggestion
- Add ``Alarm.ACKNOWLEDGED``, ``Alarm.TRIGGER``, ``Alarm.REPEAT``, and ``Alarm.DURATION`` properties
```

File: src/icalendar/alarms.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        """This is the component that contains the alarm.
+
+        This is None if you did not use Alarms.set_component().
+        """
+        return self._parent
+
+    def is_active(self) -> bool:
+        """Whether this alarm is active (True) or acknowledged (False).
+
+        E.g. in some calendar software, this is True until the user had a look
```
Comment (stevepiercy) [inline src/icalendar/alarms.py]:
```suggestion
        For example, in some calendar software, this is True until the user looks
```

File: src/icalendar/alarms.py [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+        return normalize_pytz(dt + td)
+
+    def acknowledge_until(self, dt: Optional[date]) -> None:
+        """This is the time in UTC when all the alarms of this component were acknowledged.
+
+        Only the last call counts.
+
+        Since RFC 9074 (Alarm Extension) was created later,
+        calendar implementations differ in how they acknowledge alarms.
+        E.g. Thunderbird and Google Calendar store the last time
```
Comment (stevepiercy) [inline src/icalendar/alarms.py]:
```suggestion
        For example, Thunderbird and Google Calendar store the last time
```

File: src/icalendar/alarms.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        an event has been acknowledged because of an alarm.
+        All alarms that happen before this time count as ackknowledged.
```
Comment (stevepiercy) [inline src/icalendar/alarms.py]:
```suggestion
        All alarms that happen before this time count as acknowledged.
```

File: src/icalendar/tools.py [hunk 5]
```diff
@@ -18,18 +16,58 @@ class UIDGenerator:
...
+
+def is_pytz_dt(dt: date):
+    """Whether the time requires localize() and normalize()."""
+    return is_datetime(dt) and is_pytz(dt.tzinfo)
+
+
+def normalize_pytz(dt: date):
+    """We have to normalize the time after a calculation if we use pytz.
+
+    pytz requires this funtion to be used in order to correctly calculate the
```
Comment (stevepiercy) [inline src/icalendar/tools.py]:
```suggestion
    pytz requires this function to be used in order to correctly calculate the
```

File: src/icalendar/tests/test_pytz_zoneinfo_integration.py [hunk 6]
```diff
@@ -90,3 +90,15 @@ def test_cache_is_emptied_when_tzp_is_switched(tzp, timezones, new_tzp_name):
     tzp.cache_timezone_component(timezones.pacific_fiji)
     tz2 = tzp.timezone("custom_Pacific/Fiji")
     assert tz1 is not tz2
+
+
+def test_invalid_name(tzp):
+    """Check that the provider name is ok."""
```
Comment (stevepiercy) [inline src/icalendar/tests/test_pytz_zoneinfo_integration.py]:
```suggestion
    """Check that the provider name is OK."""
```

File: src/icalendar/tests/test_issue_716_alarm_time_computation.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+Events can have alarms.
+Alarms can be in this state:
+
+- active - the user wants the alarm to pop up
+- acknowledged - the user no longer wants the alarm
+- snoozed - the user moved that alarm to another time
+
+The alarms can only work on the properties of the event like
+DTSTART, DTEND and DURATION.
```
Comment (stevepiercy) [inline src/icalendar/tests/test_issue_716_alarm_time_computation.py]:
```suggestion
DTSTART, DTEND, and DURATION.
```

File: src/icalendar/cal.py [hunk 8]
```diff
@@ -1103,6 +1228,168 @@ class Alarm(Component):
...
+    It also allows clients to track repeating alarms or alarms on recurring events or
+    to-dos to ensure that the right number of missed alarms can be tracked.
+
+    Clients SHOULD set this property to the current date-time value in UTC
+    when a calendar user acknowledges a pending alarm. Certain kinds of alarms,
+    such as email-based alerts, might not provide feedback as to when the calendar user
+    sees them. For those kinds of alarms, the client SHOULD set this property
+    when the alarm is triggered and the action is successfully carried out.
+
+    When an alarm is triggered on a client, clients can check to see if an"ACKNOWLEDGED"
```
Comment (stevepiercy) [inline src/icalendar/cal.py]:
```suggestion
    When an alarm is triggered on a client, clients can check to see if an "ACKNOWLEDGED"
```

File: src/icalendar/cal.py [hunk 9]
```diff
@@ -818,6 +923,26 @@ def duration(self) -> timedelta:
...
+
+        >>> from datetime import datetime
+        >>> from icalendar import Todo
+        >>> todo = Todo()  # empty without alarms
+        >>> todo.start = datetime(2024, 10, 26, 10, 21)
+        >>> len(todo.alarms.times)
+        0
+
+        Note that this only uses DTSTART and DUE but ignores
+        RDATE, EXDATE and RRULE properties.
```
Comment (stevepiercy) [inline src/icalendar/cal.py]:
```suggestion
        RDATE, EXDATE, and RRULE properties.
```

File: src/icalendar/cal.py [hunk 10]
```diff
@@ -619,6 +701,26 @@ class Event(Component):
...
+        >>> len(event.alarms.times)
+        1
+        >>> alarm_time = event.alarms.times[0]
+        >>> alarm_time.trigger  # The time when the alarm pops up
+        datetime.datetime(2021, 3, 2, 10, 15, tzinfo=ZoneInfo(key='America/New_York'))
+        >>> alarm_time.is_active()  # This alarm has not been acknowledged
+        True
+
+        Note that this only uses DTSTART and DTEND but ignores
+        RDATE, EXDATE and RRULE properties.
```
Comment (stevepiercy) [inline src/icalendar/cal.py]:
```suggestion
        RDATE, EXDATE, and RRULE properties.
```

File: src/icalendar/cal.py [hunk 11]
```diff
@@ -619,6 +701,26 @@ class Event(Component):
...
+        >>> event = Event.example("rfc_9074_example_1")
+        >>> len(event.alarms.times)
+        1
+        >>> alarm_time = event.alarms.times[0]
+        >>> alarm_time.trigger  # The time when the alarm pops up
+        datetime.datetime(2021, 3, 2, 10, 15, tzinfo=ZoneInfo(key='America/New_York'))
+        >>> alarm_time.is_active()  # This alarm has not been acknowledged
+        True
+
+        Note that this only uses DTSTART and DTEND but ignores
```
Comment (stevepiercy) [inline src/icalendar/cal.py]:
```suggestion
        Note that this only uses DTSTART and DTEND, but ignores
```

File: src/icalendar/cal.py [hunk 12]
```diff
@@ -818,6 +923,26 @@ def duration(self) -> timedelta:
...
+        """Compute the alarm times for this component.
+
+        >>> from datetime import datetime
+        >>> from icalendar import Todo
+        >>> todo = Todo()  # empty without alarms
+        >>> todo.start = datetime(2024, 10, 26, 10, 21)
+        >>> len(todo.alarms.times)
+        0
+
+        Note that this only uses DTSTART and DUE but ignores
```
Comment (stevepiercy) [inline src/icalendar/cal.py]:
```suggestion
        Note that this only uses DTSTART and DUE, but ignores
```

File: src/icalendar/cal.py [hunk 13]
```diff
@@ -494,11 +535,51 @@ def __eq__(self, other):
...
+
+        Note: This is analogous to the modification date and time for a
+        file in the file system.
+
+        Conformance:  This property can be specified in the "VEVENT",
+        "VTODO", "VJOURNAL", or "VTIMEZONE" calendar components.
+    """)
+
+    def is_thunderbird(self) -> bool:
+        """Whether this component has attributes that indicate that Mozilla Thunderbird createsd it."""
```
Comment (stevepiercy) [inline src/icalendar/cal.py]:
```suggestion
        """Whether this component has attributes that indicate that Mozilla Thunderbird created it."""
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
I caught only a few typos and Oxford comma issues. This is an excellent feature.

I approve once you review and make the changes.

**Comment by niccokunzmann** [comment]
@stevepiercy Thanks for reviewing the docs!

After I merging this, I will use the functionality for recurring-ical-events to calculate if an alarm happens within a time span.
I might come back with improvements in follow-up PRs.

**Comment by niccokunzmann** [comment]
Thanks for the review again! The tests pass and I added all your suggestions. Merging ...

**Pull request merged**